Filter attributes by shibmd_scope

vladimir-mencl-eresearch opened the pull request:
### All Submissions:

* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Have you added an explanation of what problem you are trying to solve with this PR?
* [X] Have you added information on what your changes do and why you chose this as your solution?
* [x] Have you written new tests for your changes?
* [X] Does your submission pass tests?
* [X] This project follows PEP8 style guide. Have you run your code against the 'flake8' linter?

Fix #297 by extending `satosa.micro_services.attribute_modifications.FilterAttributeValues` to support not only regexp filtering but also add filter types `shibmdscope_match_scope` and `shibmdscope_match_value` - equivalent to [https://shibboleth.atlassian.net/wiki/spaces/IDP4/pages/1272053894/ScopeMatchesShibMDScope](https://shibboleth.atlassian.net/wiki/spaces/IDP4/pages/1272053894/ScopeMatchesShibMDScope) and [https://shibboleth.atlassian.net/wiki/spaces/IDP4/pages/1272053892/ValueMatchesShibMDScope](https://shibboleth.atlassian.net/wiki/spaces/IDP4/pages/1272053892/ValueMatchesShibMDScope) from the Shibboleth project.

@c00kiemon5ter , please let me know whether you are happy with the overall approach - in #297, you indicated this should be done in a microservice - and in the end, it wasn't that hard.

I first thought I'd do this in a new microservice, but I'd just be replicating what `FilterAttributeValues` does - and this additional filtering fits in there conceptually well.

If happy with this approach, I'd add documentation and tests - but please let me know if overall happy with the approach I took.

Cheers,
Vlad

c00kiemon5ter commented on example/plugins/microservices/filter_attributes.yaml.example:
```diff
@@ -2,6 +2,25 @@ module: satosa.micro_services.attribute_modifications.FilterAttributeValues
+                    # enforce controlled vocabulary (via extended notation)
+                    regexp: "^(faculty|student|staff|alum|member|affiliate|employee|library-walk-in)@"
+                    # enforce correct scope
+                    shibmdscope_match_scope:
```
These filters stand here without a key.
Wouldn't it be more natural to turn them into booleans?

The code would do:
```py
                elif filter_type == "shibmdscope_match_scope" and filter_value:
```

I know this is against the implementation logic (a loop over the filters) but I think from the point of view of a user it feels better.

Let me know.

vladimir-mencl-eresearch replied:
I was considering several options for how the filters would be specified.

I wanted to:
* allow selecting the filter type (going beyond just a regexp)
* allow using multiple filters for a single attribute (eduPersonScopedAffiliation needs both the md_scope filter and a regexp filter for the value).

I considered a list - like:
```
eduPersonScopedAffiliation:
  - type: regexp
    text: staff|student
  - type: shibmdscope_match_scope
```
... but thought that would be too verbose - and the list should not be required.

So went for a dict indexed by the filter type - and meaning of value is specific to the filter type.  And some filter types do not need a value at all.

So in the syntax I picked, it is actually a key with no value (from YAML parser, with value `None`).

I think I got influenced by Ansible when picking that - Ansible uses that a lot.  Ansible tasks get specified by including a dict key matching the name of the task - and if the task needs do parameters, you just have a key with no value - like here.

I could change that - but adding any value for these keys would just make the syntax more verbose, so having just the keys seems the most concise way to me.

Your thoughts?  (Or preferences on syntax?)

c00kiemon5ter replied:
OK, let's keep it as it is now.

c00kiemon5ter commented on src/satosa/micro_services/attribute_modifications.py:
```diff
@@ -29,28 +33,62 @@ def __init__(self, config, *args, **kwargs):
+                    attributes[attribute_name] = list(filter(filter_func, attributes[attribute_name]))
+
+    def _shibmdscope_match_value(self, value, md_scopes):
+        for md_scope in md_scopes:
+            if not md_scope['regexp'] and md_scope['text'] == value:
+                return True
+            elif md_scope['regexp'] and re.compile(md_scope['text']).match(value):
+                return True
+        return False
```
Compiling the regex will not make a difference here, as the result is not reused;
I'm thinking if it would make sense to use `re.fullmatch` here, but I guess `re.match` is safer.

c00kiemon5ter commented:
@vladimir-mencl-eresearch I think it's fine! Please, go ahead 👍

vladimir-mencl-eresearch commented:
Hi @c00kiemon5ter , I've  now added tests - and also made a few tweaks, incl. removing the unnecessary re.compile - and switched to re.fullmatch as discussed.

I think it's now ready for a proper review - please let me know what you think needs to get done to get it merged.

Thanks a lot in advance for getting back to me.

Cheers,
Vlad
